src/modules/direnv.rs

```diff
@@ -0,0 +1,280 @@
+use std::borrow::Cow;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use super::{Context, Module, ModuleConfig};
+
+use crate::configs::direnv::DirenvConfig;
+use crate::formatter::StringFormatter;
+
+/// Creates a module with the current direnv rc
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("direnv");
+    let config = DirenvConfig::try_load(module.config);
+
+    let direnv_applies = !config.disabled
+        && context
+            .try_begin_scan()?
+            .set_extensions(&config.detect_extensions)
+            .set_files(&config.detect_files)
+            .set_folders(&config.detect_folders)
+            .is_match();
+
+    if !direnv_applies {
+        return None;
+    }
+
+    let direnv_status = &context.exec_cmd("direnv", &["status"])?.stdout;
```

Comment:
Just to be clear this module would be disabled by default and even when enabled will only run if a `.envrc` file is present.

In the case that the module is enabled and a `.envrc` file is present running `direnv status` is the only way to determine if the file is allowed and loaded.

Can you help clarify what other information could be provided to this module to know if it is "needed" to run?

```rust
impl<'a> Default for DirenvConfig<'a> {
    fn default() -> Self {
        Self {
            format: "[$symbol$loaded/$allowed]($style) ",
            symbol: "direnv ",
            style: "bold orange",
            disabled: true,
            detect_extensions: vec![],
            detect_files: vec![".envrc"],
            detect_folders: vec![],
            allowed_msg: "allowed",
            denied_msg: "denied",
            loaded_msg: "loaded",
            unloaded_msg: "not loaded",
        }
    }
}
```

Reply:
direnv will add some env vars when it loaded, such `$DIRENV_DIFF`, `DIRENV_DIR`, `DIRENVFILE` and `DIRENV_WATCH`. This should be helpful to do this.

Reply:
Generally speaking I think IO operations and process operations should be avoided as much as possible, these are always slowly.

Reply:
In that case this module could never provide information when a `.envrc` file is not allowed which was a request in #4962

Reply:
if .direnv file existed but  has no env vars, I think it should be  consider as "disallowed".

In fact, I don't object to you running multiple processes, I just want to be able to run it when you need it.  For example, if I just want direnv to display a symbol when it is loaded, then the direnv process should not be running

Reply:
Okay, so if I don’t plan to get the disallowed status, can I not run direnv?

Reply:
Today I re-examined the environment variables of direnv and the implementation of p10k. you're right. It is not possible to obtain the disallowed status without running direnv.
p10k checks the environment variables, which means that the icon will be displayed even in the disallowed state. I agree with your changes. But is it possible if I don't intend to get the disallowed status (like p10k does).

Reply:
I think that's a decision for the project maintainers, but I would say:

- Text parsing from the output of `direnv status` is already brittle let alone relying on environment variables used internally by `direnv` which have polluted the shell's environment.
- Using `direnv status` allows for extensibility in the future to collect and display more information via the prompt whereas the environment variables provide minimal information.
- Performance optimization without benchmarking/analysis is pre-optimization. To support your argument for deferring execution of `direnv status` I would like to see evidence that the latency incurred is statistically significant with regards to total execution time e.g. with execution of `direnv status` prompt display latency is 20% slower than without.
- This module is disabled by default so users are only affected by the additional overhead if they choose to be.

Reply:
I don't think this is a pre-optimization, this is a foreseeable optimization point. The overhead of multi-processing is typically around two milliseconds on Linux and up to 8 milliseconds on Windows. For prompt, reaching 8 milliseconds for a single component is no longer a problem that can be simply ignored. For me, the latency I've tested that I can perceive is around 80 milliseconds. Every component in a starship should have performance considerations carefully considered.

And I add a benchmark for `direnv`, here are the results:

the benchmark command is `hyperfine --warmup 1000 'direnv status' -N`

`macos m2 chip`:

```console
Benchmark 1: direnv status
  Time (mean ± σ):       4.0 ms ±   0.4 ms    [User: 1.5 ms, System: 1.2 ms]
  Range (min … max):     3.2 ms …   6.2 ms    608 runs
```

Reply:
It's pre-optimization in the sense that capabilities and design are being compromised for the sake of performance when performance has not yet been demonstrated to be an issue. Benchmarking executing `direnv` in isolation is also insufficient to understand the impact in a concurrent context especially when the primary function of the application is i/o bound i.e displaying text which is always slow compared to computation.

Reply:
Just FYI: starship executes [most things in parallel](https://github.com/starship/starship/blob/ba9f209a5fa739e6d183d896ed22550fb6a38f34/src/print.rs#L91C20-L91C20) so starships execution time is usually dominated by the biggest contributor. Looking at my timings in a git repo, thats usually git in the 20ms range:

```shell
[11:05:01] λ  starship timings

 Here are the timings of modules in your prompt (>=1ms or output):
 git_metrics    -  23ms  -   ""
 git_status     -  21ms  -   ""
 python         -  13ms  -   " v3.11.6 (venv:whatever-service-py3.11) "
 git_branch     -   3ms  -   "on  feature/som_branch "
 kubernetes     -   3ms  -   "󱃾 M3 in "
 git_commit     -   2ms  -   "(0e59472c) "
 aws            -   1ms  -   ""
 directory      -   1ms  -   "~/projects/whatever-service "
 time           -   1ms  -   "[11:17:59] "
 battery        -   1ms  -   ""
 pijul_channel  -   1ms  -   ""
 character      -  <1ms  -   "λ  "
 line_break     -  <1ms  -   "\n"
```

Reply:
It might be worth implementing the direnv allow checks ourselves, looking at the direnv config, checking `~/.local/share/direnv/allow/*` etc.
As a command that is always run, avoiding running `direnv status` seems worthwhile. But, I feel that as an initial implementation, the performance of this PR appears to be good enough for now.

Reply:
Yes, I think with iteration we can improve this. My initial concern is just that direnv does not provide a public interface without running `direnv status` so coupling to the current implementation could make this module a maintenance burden. I am definitely not certain that the existing on-disk data formats are consistent across direnv versions today so could be tricky.

For the moment I'd prefer to get it in the hands of end users to see what user feedback dictates should be prioritized.